google/cloud/forseti/scanner/scanners/external_project_access_scanner.py

```diff
@@ -0,0 +1,199 @@
+            rules (str): Fully-qualified path and filename of the rules file.
+        """
+        super(ExternalProjectAccessScanner, self).__init__(
+            global_configs,
+            scanner_configs,
+            service_config,
+            model_name,
+            snapshot_timestamp,
+            rules)
+
+        self.inventory_configs = self.service_config.get_inventory_config()
+        self.rules_engine = \
+            epa_rules_engine.ExternalProjectAccessRulesEngine(
+                rules_file_path=self.rules,
+                snapshot_timestamp=self.snapshot_timestamp
+            )
+
+        self.rules_engine.build_rule_book(self.inventory_configs)
+
+        self._ancestries = dict()
+
+    def _output_results(self, all_violations):
+        """Output results.
+
+        Args:
+            all_violations (list): A list of violations.
+        """
+        all_violations = self._flatten_violations(all_violations)
+
+        self._output_results_to_db(all_violations)
+
+    def _find_violations(self, ancestries_by_user):
+        """Find violations in the policies.
+
+        Args:
+            ancestries_by_user (dict): The project ancestries collected
+                                               from the scanner
+        Returns:
+            list: A list of ExternalProjectAccess violations
+        """
+        all_violations = []
+        LOGGER.info('Finding project access violations...')
+
+        for user_mail, project_ancestries in ancestries_by_user.iteritems():
+            for project_ancestry in project_ancestries:
+                violations = \
+                    self.rules_engine.find_policy_violations(user_mail,
+                                                             project_ancestry)
+                all_violations.extend(violations)
+
+        return all_violations
+
+    @staticmethod
+    def _flatten_violations(violations):
+        """Flatten RuleViolations into a dict for each RuleViolation member.
+
+        Args:
+            violations (list): The RuleViolations to flatten.
+
+        Yields:
+            dict: Iterator of RuleViolations as a dict per member.
+        """
+        for violation in violations:
+
+            violation_data = {
+                'full_name': violation.full_name,
+                'member': violation.member,
+                'rule_ancestor': violation.rule_ancestor.name
+            }
+
+            yield {
+                'resource_id': violation.resource_id,
+                'resource_type': violation.resource_type,
+                'full_name': violation.full_name,
+                'rule_index': violation.rule_index,
+                'rule_name': violation.rule_name,
+                'violation_type': violation.violation_type,
+                'violation_data': violation_data,
+                'resource_data': violation.resource_data
+            }
+
+    def _retrieve(self):
```

Comment:
This function is doing a lot, so it's difficult to parse all the steps. Please refactor and break out into smaller helper functions to make the flow clearer.

1. get users
2. for each user, get projects.
3. for each project, get ancestry.
4. build list of user projects
5. return dict of user -> projects

At the least, 2-4 should be put into a helper function, instead of part of this one.